src/library/scala/util/Using.scala

```diff
@@ -0,0 +1,286 @@
+package scala.util
+
+import java.util.concurrent.atomic.AtomicBoolean
+
+import scala.util.control.{ControlThrowable, NonFatal}
+
+/** A utility for performing automatic resource management. It can be used to perform an
+  * operation using resources, after which it will close the resources, in reverse order
+  * of their creation. The resource opening, operation, and resource closing are wrapped
+  * in a `Try`.
+  *
+  * If more than one exception is thrown by the operation and closing resources,
+  * the exception thrown ''first'' is returned within the `Try`, with the other exceptions
+  * [[java.lang.Throwable.addSuppressed(Throwable) added as suppressed exceptions]]
+  * to the one thrown first. This is the case ''unless'' a later exception is
+  * [[scala.util.control.NonFatal fatal]], and the one preceding it is not. In that case,
+  * the first exception is added as a suppressed exception to the fatal one, and the fatal
+  * one is thrown. If an exception is a
+  * [[scala.util.control.ControlThrowable ControlThrowable]], no exception will be added to
+  * it as a suppressed exception.
+  *
+  * @example
+  * {{{
+  * val lines: Try[List[String]] = Using(resource1) { r1 =>
+  *   r1.lines.toList
+  * }
+  * }}}
+  *
+  * @example
+  * {{{
+  * val lines: Try[Seq[String]] = for {
+  *   r1 <- Using(resource1)
+  *   r2 <- Using(resource2)
+  *   r3 <- Using(resource3)
+  *   r4 <- Using(resource4)
+  * } yield {
+  *   // use your resources here
+  *   r1.lines ++ r2.lines ++ r3.lines ++ r4.lines
+  * }
+  * }}}
+  */
+final class Using[R](resource: => R) {
+  private[this] val used = new AtomicBoolean(false)
```

Comment:
Is there a scenario where a resource will be attempted to be reclaimed in parallel?

Reply:
... not that I can think of. Honestly, it's by-name anyway, so it's probably simpler to just allow multiple creation/use

Reply:
I'm okay with switching it to just a `Boolean`. Do you consider that an important performance factor?